File: src/mod10_design_pattern/lesson11_proxy/java/ProxyLibrary.java
package mod10_design_pattern.lesson11_proxy.java;

public class ProxyLibrary implements Library {

    private RealLibrary realLibrary;
    private String userRole;

    public ProxyLibrary(String userRole) {
        this.userRole = userRole;
        realLibrary = new RealLibrary();
    }

    @Override
    public void accessBook(String bookName) {
        if (hasAccess()) {
            realLibrary.accessBook(bookName);
        } else {
            System.out.println("Access denied to the the book: " + bookName + ".Unsufficient permissions.");
        }
    }

    private boolean hasAccess() {
        return userRole.equalsIgnoreCase("student")
                || userRole.equalsIgnoreCase("professor");
    }
}

File: src/mod10_design_pattern/lesson11_proxy/kotlin/ProxyLibrary.kt
package mod10_design_pattern.lesson11_proxy.kotlin

class ProxyLibrary(private val userRole:String) : Library {

    private val realLib=RealLibrary()

    private val hasAccess:Boolean
        get() = userRole == "student" || userRole == "professor"

    override fun accessBook(bookName: String) {
        if (hasAccess){
            realLib.accessBook(bookName)
        }else{
            println("Access denied to the the book: $bookName. Unsufficient permissions.")
        }
    }
}